xwiki-platform-core/xwiki-platform-livedata/xwiki-platform-livedata-webjar/src/main/webjar/Logic.js

```diff
@@ -59,520 +67,763 @@ define(["jquery", "polyfills"], function ($) {
+            self.changeLayout(self.data.meta.defaultLayout).then(function (layout) {
+              resolve(layout);
+            }, function () {
+              reject();
+            });
+          } else {
+            console.error(err);
+            reject();
+          }
+        };
+
+        // load layout based on it's filename
+        require(["vue!" + BASE_PATH + "layouts/livedata-layout-" + layoutId + ".html"],
+          loadLayoutSuccess,
+          loadLayoutFailure
+        );
+
+      });
+    },
+
 
 
-  /**
-   * Return the property type descriptor corresponding to a property id
-   * @param {String} propertyId
-   * @returns {Object}
-   */
-  Logic.prototype.getPropertyTypeDescriptor = function (propertyId) {
-    var propertyDescriptor = this.getPropertyDescriptor(propertyId);
-    return this.data.meta.propertyTypes.find(function (typeDescr) {
-      return typeDescr.id === propertyDescriptor.type;
-    });
-  };
 
+    /**
+     * ---------------------------------------------------------------
+     * PAGINATION
+     */
 
 
-  /**
-   * Get the displayer descriptor associated to a property id
-   * @param {String} propertyId
-   */
-  Logic.prototype.getDisplayerDescriptor = function (propertyId) {
-    if (this.data.query.properties.indexOf(propertyId) === -1) { return; }
-    // property descriptor config
-    var propertyDescriptor = this.getPropertyDescriptor(propertyId);
-    var propertyDescriptorDisplayer = propertyDescriptor.displayer || {};
-    // property type descriptor config
-    var typeDescriptor = this.getPropertyTypeDescriptor(propertyId);
-    var typeDescriptorDisplayer = typeDescriptor.displayer || {};
-    // default displayer config
-    var displayerId = propertyDescriptorDisplayer.id || typeDescriptorDisplayer.id;
-    var displayer = this.data.meta.displayers.find(function (displayer) {
-      return displayer.id === displayerId;
-    });
+    /**
+     * Get total number of pages
+     * @returns {Number}
+     */
+    getPageCount: function () {
+      return Math.ceil(this.data.data.count / this.data.query.limit);
+    },
 
-    return $.extend({}, displayer, typeDescriptorDisplayer, propertyDescriptorDisplayer);
-  };
 
+    /**
+     * Get the page corresponding to the specified entry (0-based index)
+     * @param {Number} entryIndex The index of the entry. Uses current entry if undefined.
+     * @returns {Number}
+     */
+    getPageIndex: function (entryIndex) {
+      if (entryIndex === undefined) {
+        entryIndex = this.data.query.offset;
+      }
+      return Math.floor(entryIndex / this.data.query.limit);
+    },
+
+
+    /**
+     * Set page index (0-based index), then fetch new data
+     * @param {Number} pageIndex
+     * @returns {Number}
+     */
+    setPageIndex: function (pageIndex) {
+      if (pageIndex < 0 || pageIndex >= this.getPageCount()) { return; }
+      var previousPageIndex = this.data.query.offset;
```

Comment:
This doesn't look right. Maybe ``this.getPageIndex(this.data.query.offset)`` or simply ``this.getPageIndex()`` ?